lib/FrontendTool/FrontendTool.cpp

```diff
@@ -627,7 +630,45 @@ static bool performCompile(CompilerInstance &Instance,
       SF->getTypeRefinementContext()->dump(llvm::errs(), Context.SourceMgr);
     else if (Action == FrontendOptions::DumpInterfaceHash)
       SF->dumpInterfaceHash(llvm::errs());
-    else
+    else if (Action == FrontendOptions::DumpSerializedSyntaxTree) {
+      auto bufferID = SF->getBufferID();
+      assert(bufferID && "frontend should have a buffer ID "
+             "for the main source file");
+
+      // Get a full token stream with associated Trivia.
+      syntax::TokenPositionList tokens =
+        tokenizeWithTrivia(Invocation.getLangOptions(),
+                           Instance.getSourceMgr(), *bufferID);
+
+      llvm::SmallVector<Decl *, 16> topLevelDecls;
+      SF->getTopLevelDecls(topLevelDecls);
+
+      // Convert the old ASTs to the Syntax tree and print
+      // them out.
+      SyntaxASTMap ASTMap;
+      std::vector<RC<syntax::RawSyntax>> topLevelRaw;
+      for (auto *decl : topLevelDecls) {
+        if (decl->escapedFromIfConfig()) {
+          continue;
+        }
+        auto newNode = transformAST(ASTNode(decl), ASTMap,
+                                    Instance.getSourceMgr(),
+                                    *bufferID, tokens);
+        if (newNode.hasValue()) {
+          topLevelRaw.push_back(newNode->getRaw());
+        }
+      }
+
+      // Push the EOF token -- this ensures that any remaining trivia in the
+      // file is serialized as the EOF's leading trivia.
+      if (!tokens.empty() && tokens.back().first->getTokenKind() == tok::eof) {
+        topLevelRaw.push_back(tokens.back().first);
+      }
+
+      json::Output jsonOut(llvm::outs());
+      jsonOut << topLevelRaw;
+      llvm::outs() << "\n";
```

Comment:
Maybe it's just -emit-syntax or -emit-syntax-tree. Although then I'd expect it to respect -o as well.